File: patcher/core/src/main/java/me/mamiiblt/instafel/patcher/core/jobs/BuildProject.java
package me.mamiiblt.instafel.patcher.core.jobs;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.commons.io.FileUtils;
import org.apache.commons.io.filefilter.TrueFileFilter;
import org.json.JSONObject;

import brut.androlib.exceptions.AndrolibException;
import me.mamiiblt.instafel.patcher.core.source.APKSigner;
import me.mamiiblt.instafel.patcher.core.source.SourceManager;
import me.mamiiblt.instafel.patcher.core.source.SourceUtils;
import me.mamiiblt.instafel.patcher.core.source.WorkingDir;
import me.mamiiblt.instafel.patcher.core.utils.Env;
import me.mamiiblt.instafel.patcher.core.utils.Log;
import me.mamiiblt.instafel.patcher.core.utils.SmaliUtils;
import me.mamiiblt.instafel.patcher.core.utils.Utils;

public class BuildProject {

    private SmaliUtils smaliUtils = null;
    private SourceManager sourceManager = null;
    private boolean isCloneGenerated = false, isProductionMode = false;

    private List<String> appliedPatches = new ArrayList<>();
    private File buildFolder = null;
    private File cloneRefFolder = null;
    private JSONObject buildInfo = null;
    private String IG_VERSION = null, IG_VER_CODE = null, GENERATION_ID = null, IFL_VERSION = null;
    private File APK_UC = null, APK_C = null;
    private String BUILD_TS = null;

    private File APK_SIGNER_JAR = null;
    private File TESTKEY_KS = null;
    
    public void buildProject(File workingDir, String CORE_COMMIT, String PTAG, String PVERSION) throws Exception {
        Env.PROJECT_DIR = WorkingDir.getExistsWorkingDir(workingDir);
        Env.Config.setupConfig();;
        Env.Project.setupEnv();

        Log.info("Building project...");
        configureEnvironment();
        setOutputAPKFiles();
        updateInstafelEnv(CORE_COMMIT, PTAG, PVERSION);
        generateAPKs();
        signOutputs();
        generateBuildInfo(CORE_COMMIT, PTAG, PVERSION);
        Log.info("Project succesfully builded into /build folder");

        Env.Config.saveProperties();
        Env.Project.saveProperties();
    }

    private void configureEnvironment() throws IOException {
        Log.info("Initializing build environment...");
        this.BUILD_TS = String.valueOf(System.currentTimeMillis());
        this.APK_SIGNER_JAR = new File(Utils.mergePaths(Env.PROJECT_DIR,"build", "signer.jar"));
        this.TESTKEY_KS = new File(Utils.mergePaths(Env.PROJECT_DIR, "build", "testkey.keystore"));
        this.smaliUtils = new SmaliUtils(Env.PROJECT_DIR);
        this.buildFolder = new File(Utils.mergePaths(Env.PROJECT_DIR, "build"));
        if (buildFolder.exists()) {
            FileUtils.deleteDirectory(buildFolder);
            Log.info("Old build directory deleted.");
        }

        this.cloneRefFolder = new File(Env.PROJECT_DIR, "clone_ref");
        this.isCloneGenerated = cloneRefFolder.exists();
        this.isProductionMode = Env.Config.getBoolean(Env.Config.Keys.prod_mode, false);
        String[] appliedPatches = Env.Project.getString(Env.Project.Keys.APPLIED_PATCHES, "").split(",");
        for (String patch : appliedPatches) {
            this.appliedPatches.add(patch);
        }
        this.IG_VERSION = Env.Project.getString(Env.Project.Keys.INSTAGRAM_VERSION, null);
        this.IG_VER_CODE = Env.Project.getString(Env.Project.Keys.INSTAGRAM_VERSION_CODE, null);
        if (isProductionMode) {
            this.IFL_VERSION = Env.Project.getString(Env.Project.Keys.INSTAFEL_VERSION, null);
            this.GENERATION_ID = Env.Project.getString(Env.Project.Keys.GENID, null); 
                
            if (IFL_VERSION == null && GENERATION_ID == null) {
                Log.severe("Environment file is not compatible for building..");
            }
        }

        if (IG_VERSION == null && IG_VER_CODE == null) {
            Log.severe("Environment file is not compatible for building..");
        }

        sourceManager = new SourceManager();
        sourceManager.setConfig(SourceUtils.getDefaultIflConfigBuilder(sourceManager.getConfig()));
        sourceManager.getConfig().setFrameworkDirectory(SourceUtils.getDefaultFrameworkDirectory());

        Log.info("Building environment is configured succesfully");
    }

    private void signOutputs() throws Exception {
        APKSigner.moveOrDeleteApkSigner(true, APK_SIGNER_JAR, TESTKEY_KS);
        if (isCloneGenerated) {
            signAPKs(APK_UC, APK_C);
        } else {
            signAPKs(APK_UC);
        }
        APKSigner.moveOrDeleteApkSigner(false, APK_SIGNER_JAR, TESTKEY_KS);
    }

    private void signAPKs(File... APKs) throws Exception {
        Log.info("Signing APKs");
        List<String> params = new ArrayList<>();
        params.add("-a");
        for (File file : APKs){
            params.add(file.getAbsolutePath());
        }
        Log.info("Using default keystore for signing APKs");
        params.addAll(Arrays.asList(
            "--ks", TESTKEY_KS.getAbsolutePath(),
            "--ksAlias testkey",
            "--ksPass android",
            "--ksKeyPass android"
        ));   
        params.add("--overwrite");
        Log.info("Signing apks...");
        int exitCode = APKSigner.execSigner(params, APK_SIGNER_JAR);

        if (exitCode == 0) {
            Log.info("All APKs successfully signed");
        } else {
            FileUtils.deleteDirectory(buildFolder);
            Log.severe("Error while signing apk, clearing /build directory and force exiting");
        }
    }

    private void generateAPKs() throws AndrolibException, IOException {
        Log.info("Generating APKs...");
        Log.info("Generating unclone variant...");
        sourceManager.build(APK_UC.getName());
        Log.info("Unclone APK succesfully generated.");
        if (isCloneGenerated) {
            Log.info("Generating clone variant...");
            replaceSourceFiles(true);
            sourceManager.build(APK_C.getName());
            replaceSourceFiles(false);
            Log.info("Clone APK succesfully generated.");
        }
    }

    private void replaceSourceFiles(boolean type) throws IOException {
        File originalTempDir = new File(Utils.mergePaths(Env.PROJECT_DIR, "orig_temp"));
        FileUtils.forceMkdir(originalTempDir);
        Collection<File> filesAndDirs = FileUtils.listFilesAndDirs(
            cloneRefFolder,
            TrueFileFilter.TRUE,
            TrueFileFilter.TRUE
        );

        if (type) {
            for (File cloneFile : filesAndDirs) {
                if (!cloneFile.isDirectory()) {
                    File sourceFile = new File(cloneFile.getAbsolutePath().replace("clone_ref", "sources"));
                    File tempOrigFile = new File(cloneFile.getAbsolutePath().replace("clone_ref", "orig_temp"));
                    FileUtils.moveFile(sourceFile, tempOrigFile);
                    FileUtils.moveFile(cloneFile, sourceFile);
                }
            }
        } else {
            for (File cloneFile : filesAndDirs) {
                if (!cloneFile.isDirectory()) {
                    File originalFileFromTemp = new File(cloneFile.getAbsolutePath().replace("clone_ref", "orig_temp"));
                    File originalFile = new File(cloneFile.getAbsolutePath().replace("clone_ref", "sources"));
                    FileUtils.delete(originalFile);
                    FileUtils.moveFile(originalFileFromTemp, originalFile);
                }
            }
        }
    }

    private void updateInstafelEnv(String CORE_COMMIT, String PTAG, String PVERSION) throws IOException {
        if (appliedPatches.contains("copy_instafel_src")) {
            Log.info("Updating Instafel app environment...");
            String iflSourceFolder = Env.Project.getString(Env.Project.Keys.IFL_SOURCES_FOLDER, null);
            if (iflSourceFolder == null) {
                Log.severe("IFL_SOURCES_FOLDER is null");
                System.exit(-1);
            }
            File envFile = new File(
                Utils.mergePaths(Env.PROJECT_DIR, "sources", iflSourceFolder, "me", "mamiiblt", "instafel", "InstafelEnv.smali"));
            File origEnvFile = new File(Utils.mergePaths(Env.PROJECT_DIR, "sources", "InstafelEnv_orig.smali"));

            if (origEnvFile.exists()) {
                FileUtils.copyFile(origEnvFile, envFile);
            } else {
                FileUtils.copyFile(envFile, origEnvFile);
            }
                  
            List<String> fContent = smaliUtils.getSmaliFileContent(envFile.getAbsolutePath());
            Map<String, String> pairs = new HashMap<>();
            pairs.put("_iflver_", isProductionMode ? IFL_VERSION : "NOT_PROD_MODE");        
            pairs.put("_genid_", isProductionMode ? GENERATION_ID : "NOT_PROD_MODE");
            pairs.put("_igver_", IG_VERSION);
            pairs.put("_igvercode_", IG_VER_CODE);
            pairs.put("_commit_", CORE_COMMIT);
            pairs.put("_ptag", PTAG);
            pairs.put("_pversion_", "v" + PVERSION);
            pairs.put("_branch_", "main");
            String apatches = "";
            for (String patch : appliedPatches) {
                if (!patch.contains("clone")) {
                    apatches = apatches.equals("") ? apatches + patch : apatches + "," + patch;
                }
            }
            pairs.put("_patches_", apatches.trim());

            for (Map.Entry<String, String> prop : pairs.entrySet()) {
                if (prop.getValue() == null) {
                    Log.severe("Prop " + prop.getKey() + " is empty.");
                }
            }

            for (int i = 0; i < fContent.size(); i++) {
                String line = fContent.get(i);

                if (line.contains("PRODUCTION_MODE")) {
                    fContent.set(i, isProductionMode ?
                        ".field public static PRODUCTION_MODE:Z = true" :
                        ".field public static PRODUCTION_MODE:Z = false");
                }

                for (Map.Entry<String, String> prop : pairs.entrySet()) {
                    if (line.contains(prop.getKey())) {
                        fContent.set(i, line.replace(prop.getKey(), prop.getValue()));
                    }
                }
            }

            FileUtils.writeLines(envFile, fContent);
            Log.info("Instafel app environment configured succesfully");
        }
    }

    private void setOutputAPKFiles() {
        if (!isProductionMode) {
            this.APK_UC = new File(Utils.mergePaths(Env.PROJECT_DIR, "build", "unclone.apk"));
            this.APK_C = new File(Utils.mergePaths(Env.PROJECT_DIR, "build", "clone.apk"));
        } else {
            this.APK_UC = new File(Utils.mergePaths(Env.PROJECT_DIR, "build", "instafel_uc_v" + IFL_VERSION + "_" + IG_VERSION + ".apk"));
            this.APK_C = new File(Utils.mergePaths(Env.PROJECT_DIR, "build", "instafel_c_v" + IFL_VERSION + "_" + IG_VERSION + ".apk"));
        }
    }

    private void generateBuildInfo(String CORE_COMMIT, String PTAG, String PVERSION) {
        this.buildInfo = new JSONObject();
        buildInfo.put("manifest_version", 1);
        if (isProductionMode == false) {
            buildInfo.put("clone_generated", isCloneGenerated);
        }
        JSONObject patcherInfo = new JSONObject();
        patcherInfo.put("commit", CORE_COMMIT);
        patcherInfo.put("channel", PTAG);
        patcherInfo.put("version", PVERSION);        
        buildInfo.put("patcher", patcherInfo);

        JSONObject patcherData = new JSONObject();
        patcherData.put("build_date", BUILD_TS);
        if (isProductionMode) {
            JSONObject ifl = new JSONObject();
            ifl.put("version", Integer.parseInt(IFL_VERSION));
            ifl.put("gen_id", GENERATION_ID);
            patcherData.put("ifl", ifl);
        }
        JSONObject ig = new JSONObject();
        ig.put("version", IG_VERSION);
        ig.put("ver_code", IG_VER_CODE);
        patcherData.put("ig", ig);
        buildInfo.put("patcher_data", patcherData);

        if (isProductionMode) {
            JSONObject links = new JSONObject();
            links.put("unclone", "https://github.com/mamiiblt/instafel/releases/download/v" + IFL_VERSION + "/" + APK_UC.getName());
            if (isCloneGenerated) {
                links.put("clone", "https://github.com/mamiiblt/instafel/releases/download/v" + IFL_VERSION + "/" + APK_C.getName());
            }
            buildInfo.put("links", links);
        }

        JSONObject fnames = new JSONObject();
        fnames.put("unclone", APK_UC.getName());
        if (isCloneGenerated) {
            fnames.put("clone", APK_C.getName());
        }
        buildInfo.put("fnames", fnames);

        JSONObject hash = new JSONObject();
        hash.put("unclone", Utils.getFileMD5(APK_UC));
        if (isCloneGenerated) {
            hash.put("clone", Utils.getFileMD5(APK_C));
        }
        buildInfo.put("hash", hash);


        File bInfoFile = new File(Utils.mergePaths(buildFolder.getAbsolutePath(), "build_info.json"));
        
        try (FileWriter file = new FileWriter(bInfoFile)) {
            file.write(buildInfo.toString());
            Log.info("Build information file saved.");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: patcher-core/src/main/kotlin/instafel/patcher/core/jobs/BuildProject.kt
package instafel.patcher.core.jobs

import com.google.gson.Gson
import instafel.patcher.core.source.APKSigner
import instafel.patcher.core.source.SourceManager
import instafel.patcher.core.source.SourceUtils
import instafel.patcher.core.source.WorkingDir
import instafel.patcher.core.utils.Env
import instafel.patcher.core.utils.Log
import instafel.patcher.core.utils.SmaliUtils
import instafel.patcher.core.utils.Utils
import instafel.patcher.core.utils.modals.CLIJob
import instafel.patcher.core.utils.modals.pojo.BuildInfo
import instafel.patcher.core.utils.modals.pojo.FileInfo
import instafel.patcher.core.utils.modals.pojo.FileInfos
import instafel.patcher.core.utils.modals.pojo.Patcher
import instafel.patcher.core.utils.modals.pojo.PatcherData
import org.apache.commons.io.FileUtils
import org.apache.commons.io.filefilter.TrueFileFilter
import java.io.File
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import kotlin.system.exitProcess

object BuildProject: CLIJob {

    lateinit var smaliUtils: SmaliUtils
    lateinit var sourceManager: SourceManager
    lateinit var IG_VERSION: String
    lateinit var IG_VER_CODE: String
    lateinit var GENERATION_ID: String
    var IFL_VERSION = 0
    lateinit var APK_UC: File
    lateinit var APK_C: File
    lateinit var buildFolder: File
    lateinit var APK_SIGNER_JAR: File
    lateinit var TESTKEY_KS: File
    lateinit var BUILD_TS: String
    lateinit var cloneRefFolder: File

    var isCloneGenerated = false
    var isProductionMode = false

    override fun runJob(vararg args: Any) {
        val workingDir = args.getOrNull(0) as? File
        val coreCommit = args.getOrNull(1) as? String
        val projectTag = args.getOrNull(2) as? String
        val projectVersion = args.getOrNull(3) as? String

        if (workingDir !is File || coreCommit !is String || projectTag !is String || projectVersion !is String) {
            Log.severe("Wrong arguments given by CLI")
            exitProcess(-1)
        }

        Env.PROJECT_DIR = WorkingDir.getExistsWorkingDir(workingDir)
        Env.setupConfig()
        Env.setupProject()

        Log.info("Building \"${workingDir.name}\" project...")
        configureEnvironment()
        setOutputAPKFiles()
        updateInstafelEnv(coreCommit, projectTag, projectVersion)
        generateAPKs()
        signOutputs()
        generateBuildInfo(coreCommit, projectTag, projectVersion)
        Log.info("Project successfully built into /build folder of project")

        Env.saveConfig()
        Env.saveProject()
    }

    fun signOutputs() {
        APKSigner.moveOrDeleteApkSigner(true, APK_SIGNER_JAR, TESTKEY_KS)
        if (isCloneGenerated) {
            signAPKs(APK_UC, APK_C)
        } else {
            signAPKs(APK_UC)
        }
        APKSigner.moveOrDeleteApkSigner(false, APK_SIGNER_JAR, TESTKEY_KS)
    }

    fun signAPKs(vararg APKs: File) {
        Log.info("Signing APKs")

        val params = mutableListOf("-a").apply {
            addAll(APKs.map { it.absolutePath })
            Log.info("Using default keystore for signing APKs")

            addAll(
                listOf(
                    "--ks", TESTKEY_KS.absolutePath,
                    "--ksAlias", "testkey",
                    "--ksPass", "android",
                    "--ksKeyPass", "android"
                )
            )

            add("--overwrite")
        }

        Log.info("Signing apks...")
        val exitCode = APKSigner.execSigner(params, APK_SIGNER_JAR)

        if (exitCode == 0) {
            Log.info("All APKs successfully signed")
        } else {
            Log.severe("Error while signing APKs, clearing /build directory and force exiting")
            exitProcess(-1)
        }
    }

    fun generateBuildInfo(coreCommit: String, patcherTag: String, patcherVersion: String) {
        val buildInfo = BuildInfo(
            manifestVersion = 1,
            patcher = Patcher(
                version = patcherVersion,
                commit = coreCommit,
                channel = patcherTag
            ),
            patcherData = PatcherData(
                buildDate = BUILD_TS,
                generationId = GENERATION_ID,
                iflVersion = IFL_VERSION,
                igVersion = IG_VERSION,
                igVersionCode = IG_VER_CODE
            ),
            fileInfos = FileInfos(
                unclone = FileInfo(
                    fileName = APK_UC.name,
                    fileHash = Utils.getFileMD5(APK_UC)
                ),
                clone = FileInfo(
                    fileName = APK_C.name,
                    fileHash = Utils.getFileMD5(APK_C)
                )
            ),
        )

        val bInfoFile = File(Utils.mergePaths(buildFolder.absolutePath, "build_info.json"))
        bInfoFile.writeText(Env.gson.toJson(buildInfo, BuildInfo::class.java))
        Log.info("Build information file saved.")
    }


    fun generateAPKs() {
        Log.info("Generating APKs...")
        Log.info("Generating unclone variant...")
        sourceManager.build(APK_UC.name)
        Log.info("Unclone APK successfully generated...")
        if (isCloneGenerated) {
            Log.info("Generating clone variant....")
            replaceSourceFiles(true);
            sourceManager.build(APK_C.name)
            replaceSourceFiles(false);
            Log.info("Clone APK successfully generated.")
        }
    }

    fun replaceSourceFiles(type: Boolean) {
        val originalTempDir = File(Utils.mergePaths(Env.PROJECT_DIR, "orig_temp"))
        FileUtils.forceMkdir(originalTempDir)

        val filesAndDirs = FileUtils.listFilesAndDirs(
            cloneRefFolder,
            TrueFileFilter.TRUE,
            TrueFileFilter.TRUE
        )

        if (type) {
            for (cloneFile in filesAndDirs) {
                if (!cloneFile.isDirectory) {
                    val sourceFile = File(cloneFile.absolutePath.replace("clone_ref", "sources"))
                    val tempOrigFile = File(cloneFile.absolutePath.replace("clone_ref", "orig_temp"))

                    FileUtils.moveFile(sourceFile, tempOrigFile)
                    FileUtils.moveFile(cloneFile, sourceFile)
                }
            }
        } else {
            for (cloneFile in filesAndDirs) {
                if (!cloneFile.isDirectory) {
                    val originalFileFromTemp = File(cloneFile.absolutePath.replace("clone_ref", "orig_temp"))
                    val originalFile = File(cloneFile.absolutePath.replace("clone_ref", "sources"))

                    FileUtils.deleteQuietly(originalFile)
                    FileUtils.moveFile(originalFileFromTemp, originalFile)
                }
            }
        }
    }

    fun configureEnvironment() {
        Log.info("Initializing build environment...")

        smaliUtils = SmaliUtils(Env.PROJECT_DIR)
        BUILD_TS = Instant.now().toString()
        buildFolder =  File(Utils.mergePaths(Env.PROJECT_DIR, "build"))
        APK_SIGNER_JAR = File(Utils.mergePaths(Env.PROJECT_DIR, "build", "signer.jar"))
        TESTKEY_KS = File(Utils.mergePaths(Env.PROJECT_DIR, "build", "testkey.keystore"))
        cloneRefFolder = File(Env.PROJECT_DIR, "clone_ref")
        isCloneGenerated = cloneRefFolder.exists()
        isProductionMode = Env.Config.productionMode
        IG_VER_CODE = Env.Project.igVersionCode.ifEmpty { "E" }
        IG_VERSION = Env.Project.igVersion.ifEmpty { "E" }

        if (buildFolder.exists()) {
            FileUtils.deleteDirectory(buildFolder)
            Log.info("Old build directory got deleted.")
        }

        if (isProductionMode) {
            IFL_VERSION = Env.Project.iflVersion
            GENERATION_ID = Env.Project.generationId

            if (IFL_VERSION == 0 && GENERATION_ID == "E") {
                Log.severe("Environment file is not compatible for building..")
                exitProcess(-1)
            }
        } else {
            IFL_VERSION = 0
            GENERATION_ID = Env.Project.generationId.ifEmpty { "E" }
        }

        if (IG_VERSION == "E" && IG_VER_CODE == "E") {
            Log.severe("Environment file is not compatible for building..")
            exitProcess(-1)
        }

        sourceManager = SourceManager().apply {
            config = SourceUtils.getDefaultIflConfigBuilder(config)
            config.frameworkDirectory = SourceUtils.getDefaultFrameworkDirectory()
        }

        Log.info("Building environment is configured successfully")
    }

    fun setOutputAPKFiles() {
        val buildPath = Utils.mergePaths(Env.PROJECT_DIR, "build")

        if (!isProductionMode) {
            APK_UC = File(buildPath, "unclone.apk")
            APK_C = File(buildPath, "clone.apk")
        } else {
            val versionSuffix = "v${IFL_VERSION}_${IG_VERSION}"
            APK_UC = File(buildPath, "instafel_uc_$versionSuffix.apk")
            APK_C = File(buildPath, "instafel_c_$versionSuffix.apk")
        }
    }

    /**
     * Updates the InstafelEnv.smali file with build-time values.
     * 
     * This function modifies the InstafelEnv.smali file to replace placeholder values
     * with actual build information. It's important to maintain valid smali syntax:
     * - String placeholders are replaced in const-string instructions
     * - Boolean fields like PRODUCTION_MODE must be set via the <clinit> constructor,
     *   not through field declaration default values
     * 
     * File location:
     * The InstafelEnv.smali file may be located in either the primary DEX folder or the
     * iflSourceFolder depending on how the CopyInstafelSources patch executed:
     * - If primary DEX differs from the smallest smali folder, InstafelEnv.smali is moved
     *   to the primary DEX (required for early app initialization)
     * - Otherwise, it remains in the iflSourceFolder (smallest smali folder)
     * This function checks both locations to ensure the file is found regardless of the setup.
     */
    fun updateInstafelEnv(coreCommit: String, projectTag: String, patcherVersion: String) {
        var envOkCheck = false;
        Env.Project.appliedPatches.groupPatches.forEach { group ->
            group.patches.forEach { patch ->
                if (patch.shortname == "copy_instafel_src") envOkCheck = true;
            }
        }

        if (envOkCheck) {
            Log.info("Updating Instafel app environment...")

            val iflSourceFolder = Env.Project.iflSourcesFolder.ifEmpty { "E" }
            if (iflSourceFolder == "E") {
                Log.severe("IFL_SOURCES_FOLDER is not defined")
                exitProcess(-1)
            }

            // InstafelEnv.smali may be in the primary DEX or in the iflSourceFolder
            // The CopyInstafelSources patch moves it to primary DEX if they differ
            // Primary DEX is the first smali folder (usually "smali")
            val primaryDexFolder = smaliUtils.smaliFolders?.firstOrNull()
            
            val envFile = if (primaryDexFolder != null) {
                // Try primary DEX first
                val primaryDexEnvFile = File(
                    Utils.mergePaths(
                        Env.PROJECT_DIR,
                        "sources",
                        primaryDexFolder.name,
                        "instafel",
                        "app",
                        "InstafelEnv.smali"
                    )
                )
                
                if (primaryDexEnvFile.exists()) {
                    primaryDexEnvFile
                } else {
                    // Fallback to iflSourceFolder location
                    File(
                        Utils.mergePaths(
                            Env.PROJECT_DIR,
                            "sources",
                            iflSourceFolder,
                            "instafel",
                            "app",
                            "InstafelEnv.smali"
                        )
                    )
                }
            } else {
                // No primary DEX found, use iflSourceFolder
                File(
                    Utils.mergePaths(
                        Env.PROJECT_DIR,
                        "sources",
                        iflSourceFolder,
                        "instafel",
                        "app",
                        "InstafelEnv.smali"
                    )
                )
            }
            
            if (!envFile.exists()) {
                Log.severe("InstafelEnv.smali not found at ${envFile.absolutePath}")
                exitProcess(-1)
            }

            val origEnvFile = File(Utils.mergePaths(Env.PROJECT_DIR, "sources", "InstafelEnv_orig.smali"))

            if (origEnvFile.exists()) {
                FileUtils.copyFile(origEnvFile, envFile)
            } else {
                FileUtils.copyFile(envFile, origEnvFile)
            }

            val fContent = smaliUtils.getSmaliFileContent(envFile.absolutePath).toMutableList()

            val pairs = mutableMapOf(
                "_iflver_" to if (isProductionMode) IFL_VERSION.toString() else "NOT_PROD_MODE",
                "_genid_" to if (isProductionMode) GENERATION_ID else "NOT_PROD_MODE",
                "_igver_" to IG_VERSION,
                "_igvercode_" to IG_VER_CODE,
                "_commit_" to coreCommit,
                "_ptag" to projectTag,
                "_pversion_" to "v$patcherVersion",
                "_branch_" to "main"
            )

            Env.Project.appliedPatches.appliedPatchCounts["singles"] = Env.Project.appliedPatches.singlePatches.count()
            Env.Project.appliedPatches.groupPatches.forEach { group -> Env.Project.appliedPatches.appliedPatchCounts[group.shortname] = group.patches.count() }

            val gson = Gson()
            val patchesTextEscaped = gson.toJson(Env.Project.appliedPatches)
                .replace("\\", "\\\\")
                .replace("\"", "\\\"")

            pairs["_patchesjson_"] = patchesTextEscaped

            for (i in fContent.indices) {
                var line = fContent[i]

                // IMPORTANT: Only modify the PRODUCTION_MODE initialization in <clinit>, not the field declaration.
                // In smali, boolean fields cannot have default values in their declaration (unlike constant int fields in R classes).
                // The correct syntax is to modify the const/4 instruction that sets the value in the static constructor.
                // Example:
                //   .field public static PRODUCTION_MODE:Z  <- Field declaration (DO NOT MODIFY)
                //   ...
                //   const/4 v0, 0x1                         <- Sets boolean value (MODIFY THIS)
                //   sput-boolean v0, L.../PRODUCTION_MODE:Z <- Stores the value
                if (line.contains("sput-boolean") && line.contains("PRODUCTION_MODE")) {
                    // Change the boolean value being set (0x0 = false, 0x1 = true)
                    if (isProductionMode) {
                        // Ensure the previous line sets v0 to 0x1 (true)
                        if (i > 1 && fContent[i - 2].trim().startsWith("const/4")) {
                            fContent[i - 2] = "    const/4 v0, 0x1"
                        }
                    } else {
                        // Ensure the previous line sets v0 to 0x0 (false)
                        if (i > 1 && fContent[i - 2].trim().startsWith("const/4")) {
                            fContent[i - 2] = "    const/4 v0, 0x0"
                        }
                    }
                }

                for ((key, value) in pairs) {
                    if (line.contains(key)) {
                        line = line.replace(key, value)
                    }
                }

                fContent[i] = line
            }

            FileUtils.writeLines(envFile, fContent)
            Log.info("Instafel app environment configured successfully")
        }
    }
}